1.18.2 Solar Compressor

Alice0Rose opened the pull request:
Adds the Solar Compressor discussed in https://github.com/TeamPneumatic/pnc-repressurized/issues/1112.
- Generates compressed air from sunlight. While generating compressed air, it also generates heat, proportional to how much compressed air is being generated.
- The efficiency of the compressor increases with its heat up to 350­­ºC. At that point the compressed air and heat generation remain constant up to 425ºC, at which point the compressor overheats, dumping all accumulated heat and shutting down. In order to restart the compressor, a Solar Wafer has to be right clicked on it, consuming the wafer.
- The compressed air production of the compressor is roughly equivalent to the Flux Compressor. The trade-off for getting free compressed air is that the Solar Compressor only works during the day and requires a puzzle-like solution to keep it stable at max efficiency (confirmed possible with max speed upgrades using default configuration).
- [Preview video](https://youtu.be/hgoBgb2M31w)

Commit:
Adds Solar Compressor

Commit:
Merge remote-tracking branch 'origin/1.18.2' into 1.18.2_solar_heat_generator

desht commented:
This looks rather good, nice job!  I'll take a closer look at it as soon as I can (might not be for a few days due to RL commitments).

desht commented on src/generated/resources/data/pneumaticcraft/recipes/assembly/solar_wafer.json:
```diff
@@ -0,0 +1,10 @@
+{
```
I can't help feeling that this is a little cheap, for what is something quite powerful.  Though I like the idea of using the assembly line; some kind of unassembled wafer which is made from the upgrade matrix and other rare items (similar to how the netherite drill bit is crafted then assembled).  Amethyst shards perhaps?

Pressure chamber: upgrade matrix + amethyst = unassembled wafer
Assembly line (laser): unassembled wafer = solar wafer

Alice0Rose replied:
I like the idea of amethyst shards being used; to be honest I kinda forgot they existed. I think they'd make a good fill-in for the item you'd find in the NPC Bases discussed here: https://github.com/TeamPneumatic/pnc-repressurized/issues/1112#issuecomment-1374572245, if you want that idea to be implemented.

desht commented on src/main/java/me/desht/pneumaticcraft/common/block/SolarCompressorBlock.java:
```diff
@@ -0,0 +1,303 @@
+        Vec3i boundingVec3 = new Vec3i(0, 1, -1);
+
+        // Creates bounding blocks and sets their offsets
+        world.setBlock(pos.offset(boundingVec1), world.getBlockState(pos).setValue(BOUNDING, true), Block.UPDATE_ALL);
```
Uneasy about the unchecked casts here.  I know the block entity *should* always by a solar compressor, but Minecraft has a way of defeating expectations, in my experience...

Also, this would be better done with a loop.  My suggestion would be to have a static array of the bounding offsets (probably two, one for a rotation on the X axis, one of the Z axis - see comments above about making this block rotatable). Then any operations done on the bounding positions can be done by looping over that array.

desht commented on src/main/java/me/desht/pneumaticcraft/common/block/SolarCompressorBlock.java:
```diff
@@ -0,0 +1,303 @@
+        if (state.getValue(BOUNDING)) {
+            return world.getBlockState(getMainPos(state, world, pos)).getBlock() == this;
+        } else {
+            return world.isEmptyBlock(pos.offset(0,1,0)) &&
```
The compressor isn't rotatable around the Y axis? Seems like the panels are always on the +/- Z axis, right?  I think one of the first issues players will raise is to allow it to be rotatable...

Alice0Rose replied:
I don't think it would make sense for the panels to be rotatable since they follow the sun and the sun always has the same path across the sky. The base could be rotatable to allow changing the tube I/O ports' axis though.

desht replied:
Yeah, fair point. But the base should be rotatable, definitely.

desht commented on src/main/java/me/desht/pneumaticcraft/common/block/SolarCompressorBlock.java:
```diff
@@ -0,0 +1,303 @@
+        BlockPos boundingPos2 = pos.offset(0, 1, 1);
+        BlockPos boundingPos3 = pos.offset(0, 1, -1);
+
+        if (level.getBlockState(boundingPos1).getBlock() == this) {
```
This could be done with a loop

desht commented on src/main/java/me/desht/pneumaticcraft/common/block/SolarCompressorBlock.java:
```diff
@@ -0,0 +1,303 @@
+        BlockPos boundingPos3 = mainPos.offset(0, 1, -1);
+
+        // Removes bounding blocks
+        if (world.getBlockState(boundingPos1).getBlock() == this) {
```
Loop here again

desht commented on src/main/java/me/desht/pneumaticcraft/common/block/SolarCompressorBlock.java:
```diff
@@ -0,0 +1,303 @@
+            BlockEntity be = world.getBlockEntity(pos);
+
+            // Only can repair if proper block entity type
+            if (be instanceof SolarCompressorBlockEntity) {
```
instanceof pattern matching would be nice here...

desht commented on src/main/java/me/desht/pneumaticcraft/common/block/SolarCompressorBlock.java:
```diff
@@ -0,0 +1,303 @@
+                // Only can repair if broken
+                if (((SolarCompressorBlockEntity)be).isBroken()) {
+                    // Uses wafer in main hand to repair compressor
+                    if (player.getMainHandItem().getItem() == ModItems.SOLAR_WAFER.get()) {
```
A lot of repeated code here; could use a sub-method

desht commented on src/main/java/me/desht/pneumaticcraft/client/gui/SolarCompressorScreen.java:
```diff
@@ -0,0 +1,118 @@
+            }
+
+            // Adds warning if the compressor is getting too hot
+            else if (te.getTemperature() > te.MAX_TEMPERATURE - 15) {
```
Suggestion: also have the block emit smoke particles if the temperature is approaching the danger zone?  A good indicator to players who might otherwise miss it.

Commit:
- Made solar compressor base rotatable
- Improved removal logic of SolarCompressorBlock to work with rotation changes
- Added smoke particles when solar compressor is in the warning heat zone
- Added a repair sound for fixing broken solar compressors
- Made solar compressor only repairable via main hand
- Added Solar Cell item, which takes the old role of the Solar Wafer item
- Changed Solar Wafer to a crafting component for the Solar Cell
- Gave Solar Wafer texture a slight pink tint to represent the amethyst used in crafting it

Commit:
- Created an abstract class and interface for creating blocks with bounding blocks like the Solar Compressor
- Abstracted bounding block components from solar compressor block and block entity

Alice0Rose commented:
I think I addressed everything brought up. Took a bit longer because I had to fix some issues related to making the base rotatable, such as redoing the onRemove() fallback to not trigger on rotation or other blockstate updates. I also abstracted as much of the bounding block logic as I could to make creating other blocks with bounding blocks easier in the future.

desht commented:
Merged, thanks!  Now how about a 1.19 cherry-pick? :smile:
